crates/net/ress/src/connection.rs

```diff
@@ -0,0 +1,343 @@
+                let bodies = self.on_block_bodies_request(request);
+                let response = RessProtocolMessage::block_bodies(req.request_id, bodies);
+                return OnRessMessageOutcome::Response(response.encoded());
+            }
+            RessMessage::GetBytecode(req) => {
+                let code_hash = req.message;
+                trace!(target: "ress::net::connection", peer_id = %self.peer_id, %code_hash, "serving bytecode");
+                let bytecode = self.on_bytecode_request(code_hash);
+                let response = RessProtocolMessage::bytecode(req.request_id, bytecode);
+                return OnRessMessageOutcome::Response(response.encoded());
+            }
+            RessMessage::GetWitness(req) => {
+                trace!(target: "ress::net::connection", peer_id = %self.peer_id, block_hash = %req.message, "serving witness");
+                self.on_witness_request(req);
+            }
+            RessMessage::Headers(res) => {
+                if let Some(RessPeerRequest::GetHeaders { tx, .. }) =
+                    self.inflight_requests.remove(&res.request_id)
+                {
+                    let _ = tx.send(res.message);
+                } else {
+                    self.report_bad_message();
+                }
+            }
+            RessMessage::BlockBodies(res) => {
+                if let Some(RessPeerRequest::GetBlockBodies { tx, .. }) =
+                    self.inflight_requests.remove(&res.request_id)
+                {
+                    let _ = tx.send(res.message);
+                } else {
+                    self.report_bad_message();
+                }
+            }
+            RessMessage::Bytecode(res) => {
+                if let Some(RessPeerRequest::GetBytecode { tx, .. }) =
+                    self.inflight_requests.remove(&res.request_id)
+                {
+                    let _ = tx.send(res.message);
+                } else {
+                    self.report_bad_message();
+                }
+            }
+            RessMessage::Witness(res) => {
+                if let Some(RessPeerRequest::GetWitness { tx, .. }) =
+                    self.inflight_requests.remove(&res.request_id)
+                {
+                    let _ = tx.send(res.message);
+                } else {
+                    self.report_bad_message();
+                }
+            }
+        };
+        OnRessMessageOutcome::None
+    }
+}
+
+impl<P> Stream for RessProtocolConnection<P>
+where
+    P: RessProtocolProvider + Clone + Unpin + 'static,
+{
+    type Item = BytesMut;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+
+        if !this.node_type_sent {
+            this.node_type_sent = true;
+            return Poll::Ready(Some(RessProtocolMessage::node_type(this.node_type).encoded()))
+        }
+
+        loop {
```

Comment:
this loop has a bias towards incoming commands and requests, but since these don't do much work this is probably okay.